rstudio/rstudio
# Chunk grading

### Intent

Provide a way for Academy to "grade" chunk code.

### Approach

Add two new calls to `rstudioapi` to register and unregister chunk callback functions. These are functions that return a list of html output that is displayed directly under the chunk output. The callback functions are given the chunk's label and the code that was ran to aid in any grading or processing.

I added logic to `onChunkExecCompleted` to run all callback functions that have been registered everytime chunk code has ran. Each callback can returns nothing or HTML code which is passed to the backend with the `ChunkOutputFinished` client event.

When the frontend receives this data it adds the HTML output to a new IFrame displayed under any existing output. 

The original specs can be seen in 1.1 here: **https://docs.google.com/document/d/1iJhyoEJQY4ssvY69bu36fJY8T_02KAE5MJXw9t-USQc/edit#**

Related RP in rstudioapi is here: https://github.com/rstudio/rstudioapi/pull/193

### QA Notes

This feature requires two new `rstudioapi` calls so the most up to date version should be loaded from github: `remotes::install_github("rstudio/rstudioapi")`. 

You will need to create and register a chunk callback function. I've included the one I've been using below. You'll then need to register the function through the API `rstudioapi::registerChunkCallback(mycallback)`. 

Once a callback is registered, executing any chunk will run the callback function. The easiest way to see this is to open a new R notebook and execute the chunks in the provided template. Using the callback below will look like this:

![image](https://user-images.githubusercontent.com/5323711/91774364-9aafa900-ebb6-11ea-9877-e5d95d9cabfe.png)




`mycallback <- function(chunkName, chunkCode) {
  template <- "<div id=\"abc123\">\n  <div class=\"callbackActions\">\n\n  </div>\n  <div class=\"callbackResult\">\n\n  </div>\n</div>\n\n<style>\n.callbackSuccess {\n  background-color: #ACA;\n}\n\n.callbackFailure {\n  background-color: #CAA;\n}\n</style>\n<script>\n  function handle(chunkName, chunkCode, result){\n\n    var divID = \"abc123\"; // A random value generated in the R callback\n    var el = document.querySelectorAll('#' + divID)[0];\n    var resultsEl = el.querySelectorAll('.callbackResult')[0];\n    var actionsEl = el.querySelectorAll('.callbackActions')[0];\n\n    window.doGrade = function(url){\n      var request = new XMLHttpRequest();\n      request.open('POST', url, true); // Simulate a happy response\n      // request.open('GET', 'http://httpbin.org/post', true); // Simulate a bad response\n      // request.open('GET', 'http://yoidontevenexisthere.org/post', true); // Simulate an invalid request\n      request.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded; charset=UTF-8');\n  \n      request.onload = function() {\n        if (this.status >= 200 && this.status < 400) {\n          // Success!\n          var resp = this.response;\n          resultsEl.innerHTML = '<h3>Noice!</h3><p><pre>' + \n            'I ran the chunk `' + chunkName + '` with the following code:\\n\\n```\\n' +\n            chunkCode + '\\n```\\n\\n and it all worked great!' +\n            '</pre></p><p><img src=\"https://www.unilad.co.uk/wp-content/uploads/2018/07/success-kid-original.jpg\" width=\"500px\" height=\"330px\"/></p>';\n          resultsEl.classList.add('callbackSuccess');\n          resultsEl.classList.remove('callbackFailure');\n        } else {\n        console.log(\"Setting\");\n        console.log(resultsEl);\n          // We reached our target server, but it returned an error\n          resultsEl.innerHTML = '<h3>Bad response</h3><p><pre>' + this.status + ' HTTP status</pre></p><p><img src=\"https://sayingimages.com/wp-content/uploads/so-sad-cat-meme.jpg\" /></p>';\n          resultsEl.classList.add('callbackFailure');\n          resultsEl.classList.remove('callbackSuccess');\n        }\n      };\n  \n      request.onerror = function() {\n        resultsEl.innerHTML = '<h3>I can\\'t even..</h3><p><img src=\" /></p>';\n        // There was a connection error of some sort\n        resultsEl.classList.add('callbackFailure');\n        resultsEl.classList.remove('callbackSuccess');\n      };\n\n      request.send('user=person&pwd=password&organization=place&requiredkey=key');\n    }\n      \n    actionsEl.innerHTML = '';\n    resultsEl.innerHTML = '';\n\n    if (chunkName === 'nobuttons') {\n      // Don't add anything\n    } else { \n      actionsEl.innerHTML = '<button onclick=\"doGrade(\\'http://httpbin.org/post\\')\">Success</button>' +\n        '<button onclick=\"doGrade(\\'http://httpbin.org/get\\')\">Failure</button>' + \n        '<button onclick=\"doGrade(\\'http://thiserverdoesntevenexisthereyo.co.uk\\')\">Wat</button>';\n    }\n  }\n  handle(\"chunkName\", \"my code here\", {});\n</script>"
  rand <- round(runif(1, 1e8, 1e9))
  htmlres <- gsub("abc123", paste0("chunk-output-", rand), template)
  htmlres <- gsub("chunkName", chunkName, htmlres)
  htmlres <- gsub("chunkCode", chunkCode, htmlres)
  return(list(htmlres))
}`

### To Do

Currently the callback function only accepts the chunk name and code. This will be expanded to include a result parameter and possibly others but I need to get with Jonathan to go over those pieces.
